core/src/main/scala/kafka/log/Log.scala

```diff
@@ -859,12 +863,13 @@ class Log(@volatile var dir: File,
   private def analyzeAndValidateRecords(records: MemoryRecords, isFromClient: Boolean): LogAppendInfo = {
     var shallowMessageCount = 0
     var validBytesCount = 0
-    var firstOffset = -1L
+    var firstOffset: Option[Long] = Some(-1L)
```

Comment:
Hmm... I would have expected we'd use `None`

Reply:
This is really a placeholder for "firstOffset has not been initialized yet, so go initialize when you see the first message".

Reply:
I see. Would it be a little more natural to let `firstOffset` be a `Long` here, and convert it to `Option` when we construct `LogAppendInfo`. I guess I'm a little concerned about `Some(-1)` somehow leaking into the rest of the code.